service/worker/scanner/history/workflow.go

```diff
@@ -0,0 +1,127 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"time"
+
+	"github.com/uber/cadence/common/domain"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/resource"
+
+	"go.uber.org/cadence"
+	"go.uber.org/cadence/activity"
+	"go.uber.org/cadence/client"
+	"go.uber.org/cadence/worker"
+	"go.uber.org/cadence/workflow"
+)
+
+const (
+	historyScannerWFID           = "cadence-sys-history-scanner"
+	WFTypeName                   = "cadence-sys-history-scanner-workflow"
+	tlName                       = "cadence-sys-history-scanner-tasklist-0"
+	historyScavengerActivityName = "cadence-sys-history-scanner-scvg-activity"
+	infiniteDuration             = 20 * 365 * 24 * time.Hour
+)
+
+var (
+	WFStartOptions = client.StartWorkflowOptions{
+		ID:                           historyScannerWFID,
+		TaskList:                     tlName,
+		ExecutionStartToCloseTimeout: infiniteDuration,
+		WorkflowIDReusePolicy:        client.WorkflowIDReusePolicyAllowDuplicate,
+		CronSchedule:                 "0 */12 * * *",
+	}
+	activityRetryPolicy = cadence.RetryPolicy{
+		InitialInterval:    10 * time.Second,
+		BackoffCoefficient: 1.7,
+		MaximumInterval:    5 * time.Minute,
+		ExpirationInterval: infiniteDuration,
+	}
+	activityOptions = workflow.ActivityOptions{
+		ScheduleToStartTimeout: 5 * time.Minute,
+		StartToCloseTimeout:    infiniteDuration,
+		HeartbeatTimeout:       5 * time.Minute,
+		RetryPolicy:            &activityRetryPolicy,
+	}
+)
+
+// Activities is used to pass resources to activities
+type Activities struct {
+	resource resource.Resource
+	// scannerPersistenceMaxQPS the max rate of calls to persistence
+	scannerPersistenceMaxQPS   dynamicconfig.IntPropertyFn
+	maxWorkflowRetentionInDays dynamicconfig.IntPropertyFn
+}
+
+// RegisterWorkflow registers workflow and activity with worker
+func RegisterWorkflow(w worker.Worker, resource resource.Resource, dc *dynamicconfig.Collection) {
```

Comment:
nit: Instead of passing `dc` to this function, you can define `HistoryScannerOptions` like `TaskListScannerOptions`, so that we don't need to change the signature of `worker.NewConfig`